src/relay/backend/graph_executor_codegen.cc

```diff
@@ -189,9 +192,109 @@ class GraphExecutorCodegen : public backend::MemoizedExprTranslator<std::vector<
     targets_ = targets;
   }
 
+  /*!
+   * \brief Calculate the storage required to store the type of relay.Expr
+   *
+   * \param func The relay expr for which the storage is calculated
+   */
+  int64_t CalculateRelayExprSizeBytes(const Type& expr_type) {
+    if (expr_type->IsInstance<TupleTypeNode>()) {
+      auto tuple_type = Downcast<TupleType>(expr_type);
+      int64_t size = 0;
+      for (const auto& field : tuple_type->fields) {
+        size += CalculateRelayExprSizeBytes(field);
+      }
+      return size;
+    }
+    auto tensor_type = expr_type.as<TensorTypeNode>();
+    auto shape = tensor_type->shape;
+    int num_of_elements = 1;
+    for (const auto& dim_index_expr : shape) {
+      if (dim_index_expr->IsInstance<IntImmNode>()) {
+        num_of_elements *= dim_index_expr.as<IntImmNode>()->value;
+      } else {
+        // If shape is dynamic, we cannot calculate workspace in compile time.
+        num_of_elements = 0;
+      }
+    }
+    auto element_size = tensor_type->dtype.bytes();
+    return element_size * num_of_elements;
+  }
+
+  /*!
+   * \brief Update the "main" control function's metadata
+   *
+   * \param func The main function that contains calls to relay primitive functions
+   */
+  void UpdateMainWorkspaceSize(const Function& func) {
+    std::unordered_map<int, std::unordered_map<int, int>> sid_workspace;
+    std::unordered_map<int, int> device_workspace;
+    std::unordered_map<int, int> device_io;
+    std::unordered_map<int, int> device_consts;
+
+    for (const auto& kv : storage_device_map_) {
+      auto sids = kv.second[0];
+      auto devices = kv.second[1];
+      CHECK_EQ(sids.size(), devices.size());
+      for (uint32_t i = 0; i < sids.size(); i++) {
+        sid_workspace[devices[i]][sids[i]] = 0;
+        device_io[devices[i]] = 0;
+        device_consts[devices[i]] = 0;
+      }
+    }
+
+    for (const auto& kv : storage_device_map_) {
+      auto size_bytes = CalculateRelayExprSizeBytes(kv.first->checked_type());
+      auto sids = kv.second[0];
+      auto devices = kv.second[1];
+      if (kv.first->IsInstance<ConstantNode>()) {
+        for (const auto& dev : devices) {
+          device_consts[dev] += size_bytes;
+        }
+        continue;
+      } else if (kv.first->IsInstance<VarNode>() || kv.first == func->body) {
+        for (const auto& dev : devices) {
+          device_io[dev] += size_bytes;
+        }
+        continue;
+      }
+      for (uint32_t i = 0; i < sids.size(); i++) {
+        if (size_bytes > sid_workspace[devices[i]][sids[i]]) {
```

Comment:
can you add a brief note why you want the max here?

Reply:
Done